File: 2023/day01/day01.py
#!/usr/bin/env python3

import sys

input_filename = "input.txt"
if len(sys.argv) > 1 and sys.argv[1]:
    input_filename = sys.argv[1]

data = open(input_filename).read().strip()
lines = data.splitlines()

# What is the sum of all of the calibration values?

p1, p2 = 0, 0
for line in lines:
    np1, np2 = "", ""
    for i, c in enumerate(line):
        if c.isdigit():
            np1 += c
            np2 += c
        else:
            for n, num in enumerate(['zero', 'one', 'two', 'three', 'four', 'five', 'six', 'seven', 'eight', 'nine']):
                if line[i:].startswith(num):
                    np2 += str(n)

    p1 += int(np1[0] + np1[-1])
    p2 += int(np2[0] + np2[-1])

# Part One:
#
print("  Part One:", p1)

# Part Two:
# 
print("  Part Two:", p2)

File: 2023/day01/day01.cpp
// day01.cpp
//

#include <fstream>
#include <iostream>
#include <string>

#include <bits/stdc++.h>

int main(int argc, char *argv[])
{
    using namespace std;

    string filename("input.txt");
    if (argc > 1)
    {
        filename = argv[1];
    }

    ifstream file { filename };
    if (!file)
    {
        cerr << "!error: " << filename << " could not be opened for reading!\n";
        return 1;
    }

    vector<string> lines;
    string line;
    while (getline(file, line))
    {
        lines.push_back(line);
    }

    int p1 = 0;
    for (string line : lines)
    {
        vector<int> v;
        for (char c : line)
        {
            if (isdigit(c))
            {
                v.push_back(c - '0');
            }
        }

        p1 += v.front() * 10 + v.back();
    }

    cout << "  Part One: " << p1 << '\n';

    int p2 = 0;
    vector<string> numbers { "zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine" };
    for (string line : lines)
    {
        vector<int> v;

        for (int i = 0; i < line.length(); ++i)
        {
            char c = line[i];
            if (isdigit(c))
            {
                v.push_back(c - '0');
            }
            else
            {
                for (int n = 0; n < numbers.size(); ++n)
                {
                    if (i + numbers[n].length() > line.length())
                        continue;

                    if (line.substr(i, numbers[n].length()) == numbers[n])
                    {
                        v.push_back(n);
                        break;
                    }
                }
            }
        }

        p2 += v.front() * 10 + v.back();
    }

    cout << "  Part Two: " << p2 << '\n';

    return 0;
}
